pyiceberg/io/pyarrow.py

```diff
@@ -288,6 +288,8 @@ def create(self, overwrite: bool = False) -> OutputStream:
         try:
             if not overwrite and self.exists() is True:
                 raise FileExistsError(f"Cannot create file, already exists: {self.location}")
+            # Parent directories must be created first in certain file systems, such as the LocalFileSystem.
+            self._filesystem.create_dir(os.path.dirname(self._path), recursive=True)
```

Comment:
This is typically something that we try to avoid. Iceberg is designed to work with object stores, and those don't have a notion of directories. One recommendation is even to disallow moves and listing of directories. One thing is also creating a directory. I'm not sure what the behavior is for the Arrow S3 implementation. Since some of the implementations still make a call(s):

- For example, a list operation to check if the directory is there
- For example, they touch a small file under the prefix to indicate that the path should be created.

Another concern is that we currently do this in Arrow, we also would need to do this for other implementations to avoid discrepancies. The concept of the FileIO is that you easily can swap them out.

Reply:
@Fokko thanks for the review. 

I agree with the above. The Arrow FileIO implementation might not be the best place to implement this behavior. So far both of the supported FS implementations (`ARROW_FILE_IO` and `FSSPEC_FILE_IO`) are failing to write to the local file system. 

I want to make writes work for the local file system. 

Looking at the Java side, there is a [`LocalOutputFile` implementation](https://github.com/apache/iceberg/blob/fd1cf49280bde07d67c6bc1a6ec60238e1e38f7f/api/src/main/java/org/apache/iceberg/Files.java#L59) which implements the behavior for creating parent directories.

Maybe we can implement a new FileIO implementation and make that the preferred implementation for the `file://` scheme.

https://github.com/apache/iceberg-python/blob/4cf1f35dfd3e7cfb2996887e861d740239746306/pyiceberg/io/__init__.py#L278

Reply:
Hey @kevinjqliu That sounds like a reasonable alternative to me 👍

Reply:
Thinking a bit more about it, maybe we should still use the ArrowFileSystem in this new FileIO to make sure that we can read the table into a Arrow `pa.Table`.

Reply:
@Fokko I updated the PR for both PyArrow and FsSpec.